### PR TITLE
feat!: Use pandas custom data types for BigQuery DATE and TIME columns, remove `date_as_object` argument

### LINKED ISSUE #861: remove `date_as_object` argument from `to_dataframe`
I believe this was added to support https://github.com/googleapis/python-bigquery/issues/11 and https://github.com/googleapis/python-bigquery/issues/136, but I'm not convinced that this was the right approach. Instead, what we do for `TIMESTAMP` columns seems more appropriate -- try to cast to the right data type. If it fails due to out-of-range, then fallback to objects.

### LINKED ISSUE #862: convert time columns to dbtime by default in `to_dataframe`
Currently TIME columns are just exposed as string objects. This would be a better experience and align with better with the expectations for working with timeseries in pandas https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html

Presumably one could combine a date column with a time column to create a datetime by adding them.

### PR DESCRIPTION
**Pull request by jimfulton**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixed #861 
Fixes #862 🦕

BREAKING CHANGE: remove `date_as_object` argument from `to_dataframe`. The `dbdate` dtype is used by default with an automatic fallback to `object` when dates are not within the range of a nanosecond-precision pandas timestamp

### CODE REVIEW

**Comment by jimfulton** [comment]
supersedes #959

File: google/cloud/bigquery/dtypes.py
Comment (tswast) [inline google/cloud/bigquery/dtypes.py:390]: do we define this or does it come from pandas? If we do, we usually add a `_` suffix to anything that has the same name as a built-in function. `type_`.

Comment (jimfulton) [reply google/cloud/bigquery/dtypes.py:390]: From pandas
https://pandas.pydata.org/docs/reference/api/pandas.api.extensions.ExtensionDtype.html

File: google/cloud/bigquery/dtypes.py
Comment (plamut) [inline google/cloud/bigquery/dtypes.py:360]: (nit) There are a bit too many zeros to parse this easily, let's add separators.
```suggestion
                1970, 1, 1, int(h), int(m), int(s), int(us * 1_000_000)
```

File: google/cloud/bigquery/dtypes.py
Comment (plamut) [inline google/cloud/bigquery/dtypes.py:357]: Is ISO string guaranteed or assumed here? Asking because a potential _"not enough / to many values to unpack"_ would not be that informative.

Is this taken care of somewhere in Pandas?

File: samples/snippets/pandas_date_and_time.py
Comment (plamut) [inline samples/snippets/pandas_date_and_time.py:17]: Is black formatting too invasive here? Re-formatting multiline `pd.Series(...)` snippets is IMO not that bad, or is there something else?

(might be subjective, don't put too much weight on it)

File: testing/constraints-3.7.txt
Comment (plamut) [inline testing/constraints-3.7.txt:1]: I see that `pandas==1.1.0` still supports Python 3.6, thus presumably this constraint is to test dtypes-related things that we use here and are not necessarily available in less recent pandas version, forcing to use our own forward ported versions?

It would probably be good to explain in a comment what feature differences do the cutoff points 1.1.0 and 1.2.0 (below) represent.

**Review by plamut — COMMENTED** [review]
This looks pretty cool, the comments are minor mostly hygiene concerns / readability suggestions.

As for the forward ported code in dtypes: It seems good, although I'm not familiar with pandas internals to say for sure.

**Comment by jimfulton** [comment]
Thanks for the comments. I'll respond this weekend. :)

**Comment by tswast** [comment]
@plamut @jimfulton I've gone through the necessary steps to create https://github.com/googleapis/python-db-dtypes-pandas from the `dtypes` module in this PR. I propose we update this PR to use that package (I'll work on cutting a pre-release today) and send the requested cleanup PRs there.

Thanks!

**Comment by tswast** [comment]
I've uploaded a pre-release to https://pypi.org/project/db-dtypes/ so that this PR can be updated to use that package while we prep that repo for a full proper release.

File: google/cloud/bigquery/dtypes.py
Comment (tswast) [inline google/cloud/bigquery/dtypes.py:35]: The `.release` attribute was added in packaging 17.0. We'll have to bump the minimum in `setup.py` (and add a constraint, since it seems we missed this)

https://github.com/pypa/packaging/blob/e2bc8f6ef2d66f6eaaf522fcf70a0f203b734064/CHANGELOG.rst#170---2017-02-28

I'm addressing this in https://github.com/googleapis/python-db-dtypes-pandas/pull/1

**Comment by jimfulton** [comment]
> @plamut @jimfulton I've gone through the necessary steps to create https://github.com/googleapis/python-db-dtypes-pandas from the `dtypes` module in this PR.

Way cool!

> I propose we update this PR to use that package (I'll work on cutting a pre-release today) and send the requested cleanup PRs there.

+1

Jim

File: google/cloud/bigquery/dtypes.py
Comment (jimfulton) [reply google/cloud/bigquery/dtypes.py:48]: Addressed here: https://github.com/googleapis/python-db-dtypes-pandas/pull/9/commits/03a18397bbecb61b7fc57af0df59b40e7cdd2c21

Comment (jimfulton) [reply google/cloud/bigquery/dtypes.py:48]: Thanks :)

File: google/cloud/bigquery/dtypes.py
Comment (jimfulton) [reply google/cloud/bigquery/dtypes.py:306]: Yeah.  We did this with sqlalchemy too, because these are complex apis that change between versions. :/

File: google/cloud/bigquery/dtypes.py
Comment (jimfulton) [reply google/cloud/bigquery/dtypes.py:360]: https://github.com/googleapis/python-db-dtypes-pandas/pull/9/commits/88a87fc8440702bdc8ced7a7e4bd93a02c8fa1fc

File: samples/snippets/pandas_date_and_time.py
Comment (jimfulton) [reply samples/snippets/pandas_date_and_time.py:17]: These examples show up in HTML pages where there isn't much width.  Blacken makes examples so wide that readers have to scroll horizontally, which is mean. :)

File: google/cloud/bigquery/dtypes.py
Comment (jimfulton) [reply google/cloud/bigquery/dtypes.py:357]: ooooooookaaaaaay

https://github.com/googleapis/python-db-dtypes-pandas/pull/9/commits/c93dd442477f0dd1860651fed7f5e2238fc3e625

File: testing/constraints-3.7.txt
Comment (jimfulton) [reply testing/constraints-3.7.txt:1]: https://github.com/googleapis/python-db-dtypes-pandas/pull/9/commits/57cfc2e9c487a71a1f09482f07f9f7701c85f282

**Comment by jimfulton** [comment]
@tswast I addressed @plamut's comments in the other repo as part of: https://github.com/googleapis/python-db-dtypes-pandas/pull/9

**Comment by jimfulton** [comment]
> I've uploaded a pre-release to https://pypi.org/project/db-dtypes/ so that this PR can be updated to use that package while we prep that repo for a full proper release.

Done

File: samples/snippets/pandas_date_and_time.py
Comment (plamut) [reply samples/snippets/pandas_date_and_time.py:17]: Ah, OK, that's indeed terrible much. :)

**Comment by tswast** [comment]
@googlebot I consent.

File: google/cloud/bigquery/table.py
Comment (tswast) [inline google/cloud/bigquery/table.py:1953]: Rather than a special check for the date dtype, I wonder if we can use the `types_mapper` parameter on `to_pandas`?

https://arrow.apache.org/docs/python/generated/pyarrow.RecordBatch.html#pyarrow.RecordBatch.to_pandas

File: samples/geography/requirements.txt
Comment (tswast) [inline samples/geography/requirements.txt:10]: needs update to 0.2.0

File: samples/snippets/requirements.txt
Comment (tswast) [inline samples/snippets/requirements.txt:1]: needs update to 0.2.0

**Review by tswast — CHANGES_REQUESTED** [review]

File: google/cloud/bigquery/table.py
Comment (tswast) [reply google/cloud/bigquery/table.py:1953]: Yes, it should be possible, but we'll need some additional work on the db-dtypes package. https://github.com/googleapis/python-db-dtypes-pandas/issues/38

**Review by tswast — APPROVED** [review]

File: tests/unit/job/test_query_pandas.py
Comment (leahecole) [inline tests/unit/job/test_query_pandas.py:126]:
```suggestion
        ["John", "Paul", "George", "Ringo"], type=pyarrow.string()
```

want to remove religious references to play it safe around code of conduct

File: tests/unit/job/test_query_pandas.py
Comment (leahecole) [inline tests/unit/job/test_query_pandas.py:523]:
```suggestion
        ["John", "Paul", "George", "Ringo"], type=pyarrow.string()
```

want to remove religious references to play it safe around code of conduct

**Review by leahecole — DISMISSED** [review]
partway through looking but did want to get this one small change in fairly early

**Review by leahecole — COMMENTED** [review]
lgtm but I don't know if I'm the most qualified reviewer for this - I didn't see anything that looked glaringly like it would break and the small changes I requested were addressed so I'm happy 🤷🏻‍♀️

**Comment by tswast** [comment]
Thanks @leahecole ! I'll let Kokoro run and check back on this PR with fresh eyes tomorrow before merging.

**Pull request merged**
